activesupport/lib/active_support/duration.rb

```
@@ -113,19 +116,49 @@ def ago(time = ::Time.current)
     def inspect #:nodoc:
       parts.
         reduce(::Hash.new(0)) { |h,(l,r)| h[l] += r; h }.
-        sort_by {|unit,  _ | [:years, :months, :days, :minutes, :seconds].index(unit)}.
+        sort_by {|unit,  _ | [:years, :months, :weeks, :days, :hours, :minutes, :seconds].index(unit)}.
         map     {|unit, val| "#{val} #{val == 1 ? unit.to_s.chop : unit.to_s}"}.
         to_sentence(locale: ::I18n.default_locale)
     end
 
+    # JSON representation should be a string formatted as ISO 8601 Duration, as recommended by Google JSON Style Guide:
+    # https://google-styleguide.googlecode.com/svn/trunk/jsoncstyleguide.xml?showone=Time_Duration_Property_Values#Time_Duration_Property_Values
     def as_json(options = nil) #:nodoc:
-      to_i
+      iso8601(precision: options && options[:precision])
     end
 
     def respond_to_missing?(method, include_private=false) #:nodoc:
       @value.respond_to?(method, include_private)
     end
 
+    # Creates a new Duration from string formatted according to ISO 8601 Duration.
+    #
+    # See http://en.wikipedia.org/wiki/ISO_8601#Durations
+    # This method allows negative parts to be present in pattern.
+    # If invalid string is provided, it will raise +ActiveSupport::Duration::ISO8601Parser::ParsingError+.
+    def self.parse!(iso8601duration)
+      parts = ISO8601Parser.new(iso8601duration).parse!
+      time  = ::Time.current
+      new(time.advance(parts) - time, parts)
+    end
+
+    # Creates a new Duration from string formatted according to ISO 8601 Duration.
+    #
+    # See http://en.wikipedia.org/wiki/ISO_8601#Durations
+    # This method allows negative parts to be present in pattern.
+    # If invalid string is provided, nil will be returned.
+    def self.parse(iso8601duration)
+      parse!(iso8601duration)
+    rescue ISO8601Parser::ParsingError
+      nil
+    end
```

IMO we don't need a method that swallows the parse error.